### ISSUE TITLE
Big IP ASM support request

### ISSUE THREAD

**Issue by jkim711** [F5Networks/f5-ansible#188]
##### ISSUE TYPE

 - Feature Idea

##### ANSIBLE VERSION

```
    Any version
```

##### PYTHON VERSION

```
   Any Version
```

##### BIGIP VERSION

```
  BIGIP ASM (Web firewall)
```

##### OS / ENVIRONMENT

N/A
##### SUMMARY

Request for Ansible support in BIGIP ASM
##### STEPS TO REPRODUCE

  Currently BIGIP ASM support REST API interface, we woud like to have an equivalent feature that uses Ansible

```

```

##### EXPECTED RESULTS

  Be able to control and configure the BIGIP ASM
##### ACTUAL RESULTS

 

```

```

**Comment by caphrim007** [comment]
@jkim711 this is a pretty broad request. Can you provide some more details on what exactly you are looking for from ASM support?

I understand "everything", but so that I can start somewhere, it would be helpful to know a specific feature or set of features you're interested in.

**Comment by jkim711** [comment]
[F5 -ASM Automatic Installation basedon Rest API Guide.pptx](https://github.com/F5Networks/f5-ansible/files/995275/F5.-ASM.Automatic.Installation.basedon.Rest.API.Guide.pptx)

This is the document we received from F5 Korea. They provided us the features through REST API. We have requested that these equivalent features be made available through Ansible.

I have feeling that many of these API are already piece-bit available in existing Ansible modules.

Thanks

**Comment by caphrim007** [comment]
thanks for that @jkim711. this is going to take some time, but I will use that ppt as a reference. I think Wojciech did a lot of work on the ASM stuff already in the F5 sdk that can be used.

**Comment by jkim711** [comment]
How can I get the F5 SDK?

**Comment by caphrim007** [comment]
@jkim711 it's found here

https://github.com/F5Networks/f5-common-python

and is installable via pip.

`pip install f5-sdk`

**Comment by jkim711** [comment]
Can this SDK be used through Ansible?

**Comment by caphrim007** [comment]
it _is_ used through Ansible. It's the foundation of all the modules that F5 writes.

**Comment by jkim711** [comment]
So if I want to create a Ansible playbook for ASM, I would need to create an Ansible module that uses SDK.

**Comment by caphrim007** [comment]
you don't need to get that complicated unless you want to.

you can always write a python script to do what you want and use the `script` or `command` modules with `delegate_to: localhost` to run arbitrary scripts.

Modules are a more formal way of doing the above.

**Comment by jkim711** [comment]
But we were asked to use Ansible. What would be way to do this?

**Comment by caphrim007** [comment]
you can use ansible. a python script can be run in ansible. an ansible module is just a python script with some extra ansible includes, but even those includes are optional.

If you want a set of ASM modules from F5, then that will take some time.

**Comment by mcgonagle** [comment]
Running an ASM python script via Ansible is a great interim recommendation.

**Comment by jkim711** [comment]
Where can I get sample code for using ASM with F5 SDK
I ran the following code.
************************************************
from f5.bigip import BigIP
import requests
from requests.packages.urllib3.exceptions import InsecureRequestWarning

requests.packages.urllib3.disable_warnings(InsecureRequestWarning)

mgmt = BigIP("192.168.199.241", "admin", "admin")
# Get a list of all pools on the BigIP and print their names and their
# members' names
pools = mgmt.tm.ltm.pools.get_collection()
for pool in pools:
    print pool.name
    for member in pool.members_s.get_collection():
         print member.name

# Create a new pool on the BIG-IP
mypool = mgmt.tm.ltm.pools.pool.create(name='mypool', partition='Common')

# Load an existing pool and update its description
pool_a = mgmt.tm.ltm.pools.pool.load(name='mypool', partition='Common')
pool_a.description = "New description"
pool_a.update()

# Delete a pool if it exists
if mgmt.tm.ltm.pools.pool.exists(name='mypool', partition='Common'):
    pool_b = mgmt.tm.ltm.pools.pool.load(name='mypool', partition='Common')
    pool_b.delete()
******************************************
I got the following result.
**************************************
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    pools = mgmt.tm.ltm.pools.get_collection()
  File "/usr/lib/python2.7/site-packages/f5/bigip/mixins.py", line 102, in __getattr__
    raise AttributeError(error_message)
AttributeError: '<class 'f5.bigip.BigIP'>' object has no attribute 'tm'

**Comment by wojtek0806** [comment]
Please do not use BigIp fixture its deprecated and will be removed at some point. Use ManagementRoot instead, from f5.bigip import ManagementRoot.

**Comment by jkim711** [comment]
I executed the following code
import requests
from requests.packages.urllib3.exceptions import InsecureRequestWarning

requests.packages.urllib3.disable_warnings(InsecureRequestWarning)

from f5.bigip import ManagementRoot

mgmt = ManagementRoot("70.60.207.94", "admin", "admin")

collection = mgmt.tm.asm.get_collection()

for anobject in collection:

print anobject
p_collections = mgmt.tm.asm.policies_s.get_collection() 
for p_object in p_collections:

print p_object.name
print p_object.id
print p_object.kind
print p_object.selfLink
print p_object.allowedResponseCodes
policy=mgmt.tm.asm.policies_s.policy.load(id=p_object.id)
I got the following results

{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/policies?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/signatures?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/violations?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/signature-statuses?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/refinements?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/signature-systems?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/file-transfer?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/attack-types?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/sub-violations?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/tasks?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/scraping-extensions?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/signature-update?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/search-engines?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/policy-templates?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/http-methods?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/signature-sets?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/events?ver=12.1.0'}}
{u'reference': {u'link': u'https://localhost/mgmt/tm/asm/file-content-detection-formats?ver=12.1.0'}}
rest-api-07
pwlpR2LJzxplAjAR3lp10g
tm:asm:policies:policystate
https://localhost/mgmt/tm/asm/policies/pwlpR2LJzxplAjAR3lp10g?ver=12.1.0
[400, 401, 404, 407, 417, 503]

All is good so far. 

The following REST API would pass the Request data for arguments.
I am trying to figure out how to the equivalent (pass arguments) the using using the F5 SDK

https://{{big_ip_a_mgmt}}/mgmt/tm/asm/policies?ver=12.1.0
Request Data
 { 
"fullPath":"/Common/rest-api_12",
"active":"true",
"virtualServers":[],
"enforcementMode":
"transparent",
"description":"none",
"applicationLanguage":"utf-8",
"policy-builder":"disabled",
"learningMode":"manual",
"stagingSettings": {
        "signatureStaging": true,
        "placeSignaturesInStaging": false,
        "enforcementReadinessPeriod": 7
      }

**Comment by caphrim007** [comment]
@jkim711 you can change settings by just passing the names of the attributes you want to change as keyword arguments to a resource that you have loaded.

```
policy=mgmt.tm.asm.policies_s.policy.load(id=p_object.id)
policy.update(description="bar", learningMode="automatic")
```

**Comment by jkim711** [comment]
I got the following results.

---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
/home/jaekim711/f5/test2.py in <module>()
     21     print p_object.allowedResponseCodes
     22     policy=mgmt.tm.asm.policies_s.policy.load(id=p_object.id)
---> 23     policy.update(description="bar", learningMode="automatic")
     24 
     25 

/usr/lib/python2.7/site-packages/f5/bigip/resource.pyc in update(self, **kwargs)
   1286         """
   1287         raise UnsupportedOperation(
-> 1288             "%s does not support the update method" % self.__class__.__name__
   1289         )
   1290

**Comment by wojtek0806** [comment]
Update method uses HTTP PUT and it is not supported on any ASM endpoint in the API, you would have to use modify() method, which uses HTTP PATCH, in this case this would be something like 

```
policy.modify(description="bar", learningMode="automatic").
```

**Comment by jkim711** [comment]
How are the results checked?
I was looking  through https://github.com/F5Networks/f5-common-python/blob/development/f5/bigip/tm/asm/test/functional/test_policies.py
But I couldn't find any hints.

**Comment by caphrim007** [comment]
@jkim711 i'm not sure what you mean by "how are the results checked". can you rephrase the question?

**Comment by jkim711** [comment]
How do I know if the modify was successful. And if there are errors, how do I fetch them?

**Comment by caphrim007** [comment]
@jkim711 if the modify fails, the method will raise an `F5SDKError` error which you can import like so

```
from f5.sdk_exceptions import F5SDKError
```

and you can fetch them as you would any other python exception

```
try:
     ...
except F5SDKError as e:
    print(str(e))
```

**Comment by caphrim007** [comment]
referencing #80 for future work

**Comment by caphrim007** [comment]
#280  and #279 are also part of this effort

**Comment by caphrim007** [comment]
the pr for this has been merged into this repo. unit tests and functional tests are also available.

**Comment by caphrim007** [comment]
##### ISSUE TYPE

 - Upstream request

##### COMPONENT NAME


##### ANSIBLE VERSION

```

```

##### PYTHON VERSION

```

```

##### TESTED VERSIONS

**Python 2.7.14**
- [ ] BIG-IP 12.0.0 (BIGIP-12.0.0.0.0.606)
- [ ] BIG-IP 12.1.0 (BIGIP-12.1.0.0.0.1434)
- [ ] BIG-IP 12.1.0-hf1 (BIGIP-12.1.0.1.0.1447-HF1)
- [ ] BIG-IP 12.1.0-hf2 (BIGIP-12.1.0.2.0.1468-HF2)
- [ ] BIG-IP 12.1.1 (BIGIP-12.1.1.0.0.184)
- [ ] BIG-IP 12.1.1-hf1 (BIGIP-12.1.1.1.0.196-HF1)
- [ ] BIG-IP 12.1.1-hf2 (BIGIP-12.1.1.2.0.204-HF2)
- [ ] BIG-IP 12.1.2 (BIGIP-12.1.2.0.0.249)
- [ ] BIG-IP 12.1.2-hf1 (BIGIP-12.1.2.1.0.264-HF1)
- [ ] BIG-IP 13.0.0 (BIGIP-13.0.0.0.0.1645)
- [ ] BIG-IP 13.0.0-hf1 (BIGIP-13.0.0.1.0.1668-HF1)
- [ ] BIG-IP 13.0.0-hf2 (BIGIP-13.0.0.2.0.1671-HF2)

**Python 3.5.4**
- [ ] BIG-IP 12.0.0 (BIGIP-12.0.0.0.0.606)
- [ ] BIG-IP 12.1.0 (BIGIP-12.1.0.0.0.1434)
- [ ] BIG-IP 12.1.0-hf1 (BIGIP-12.1.0.1.0.1447-HF1)
- [ ] BIG-IP 12.1.0-hf2 (BIGIP-12.1.0.2.0.1468-HF2)
- [ ] BIG-IP 12.1.1 (BIGIP-12.1.1.0.0.184)
- [ ] BIG-IP 12.1.1-hf1 (BIGIP-12.1.1.1.0.196-HF1)
- [ ] BIG-IP 12.1.1-hf2 (BIGIP-12.1.1.2.0.204-HF2)
- [ ] BIG-IP 12.1.2 (BIGIP-12.1.2.0.0.249)
- [ ] BIG-IP 12.1.2-hf1 (BIGIP-12.1.2.1.0.264-HF1)
- [ ] BIG-IP 13.0.0 (BIGIP-13.0.0.0.0.1645)
- [ ] BIG-IP 13.0.0-hf1 (BIGIP-13.0.0.1.0.1668-HF1)
- [ ] BIG-IP 13.0.0-hf2 (BIGIP-13.0.0.2.0.1671-HF2)

**Python 3.6.2**
- [ ] BIG-IP 12.0.0 (BIGIP-12.0.0.0.0.606)
- [ ] BIG-IP 12.1.0 (BIGIP-12.1.0.0.0.1434)
- [ ] BIG-IP 12.1.0-hf1 (BIGIP-12.1.0.1.0.1447-HF1)
- [ ] BIG-IP 12.1.0-hf2 (BIGIP-12.1.0.2.0.1468-HF2)
- [ ] BIG-IP 12.1.1 (BIGIP-12.1.1.0.0.184)
- [ ] BIG-IP 12.1.1-hf1 (BIGIP-12.1.1.1.0.196-HF1)
- [ ] BIG-IP 12.1.1-hf2 (BIGIP-12.1.1.2.0.204-HF2)
- [ ] BIG-IP 12.1.2 (BIGIP-12.1.2.0.0.249)
- [ ] BIG-IP 12.1.2-hf1 (BIGIP-12.1.2.1.0.264-HF1)
- [ ] BIG-IP 13.0.0 (BIGIP-13.0.0.0.0.1645)
- [ ] BIG-IP 13.0.0-hf1 (BIGIP-13.0.0.1.0.1668-HF1)
- [ ] BIG-IP 13.0.0-hf2 (BIGIP-13.0.0.2.0.1671-HF2)

##### CHECKLIST

#### Module code related
- [ ] Module code has `EXAMPLES`
- [ ] Module code has `DOCUMENTATION`
- [ ] Module code has `RETURN`
- [ ] Module code has `ANSIBLE_METADATA`
- [ ] Module code conforms to coding standards (v3). This includes a `Parameters` class, `ModuleManager` class, and `ArgumentSpec` class.
- [ ] Module includes GPL3 license at top of the file and F5 Networks copyright.
- [ ] Module imports are not `from *`
- [ ] If this module is new to Ansible, the `version_added` key of the `DOCUMENTATION` variable is set to `2.5`.
- [ ] If this module is not new to Ansible, but you have added accepted parameters to it, the `version_added` key for each of those new parameters is `2.5`.

#### Integration test related
- [ ] Playbook exists to do functional tests
- [ ] Role directory with functional tests exists
- [ ] Module code has been tested on major releases post 12.x. If a major release cannot be supported, document "why" in the `notes` section of the module 
`DOCUMENTATION`

#### Unit test related
- [ ] Unit tests in `tests/unit` for the product exist
- [ ] `TestParameters` unit test class exists and has content
- [ ] `TestManager` unit test class (or `*Managers` if required) exists and has content

#### QC related
- [ ] Module code passes `pycodestyle library/MODULE_NAME.py`
- [ ] Unit test code passes `pycodestyle test/unit/PRODUCT/test_MODULE_NAME.py`

#### Docs related
- [ ] Docs exist and build correctly (`make docs`)
- [ ] Docs have been merged to `f5-ansible` so they can be rebuilt by ReadTheDocs.

#### Upstream vendor related, with f5-sdk

This must be done inside of Ansible's source tree. (`cd local/ansible`)

- [ ] Module code in Ansible repo passes `ansible-test units --python 2.7 MODULE_NAME`
- [ ] Module code in Ansible repo passes `ansible-test units --python 3.5 MODULE_NAME`
- [ ] Module code in Ansible repo passes `ansible-test units --python 3.6 MODULE_NAME`
- [ ] Module code in Ansible repo passes `ansible-test sanity --test validate-modules`
- [ ] Module code in Ansible repo passes `ansible-test sanity --test pep8`

#### Upstream vendor related, without f5-sdk

This must be done inside of Ansible's source tree. (`cd local/ansible`) 

- [ ] Module code in Ansible repo passes `ansible-test sanity --test import --python 2.7`
- [ ] Module code in Ansible repo passes `ansible-test sanity --test import --python 3.5`
- [ ] Module code in Ansible repo passes `ansible-test sanity --test import --python 3.6`
- [ ] Module code in Ansible repo passes `nosetests test/units/modules/network/f5/test_MODULE_NAME.py` using venv made with requirements.bare.txt

#### Upstream vendor related, misc
- [ ] Module PR has been mentioned as a comment for the next Networking Team meeting here https://github.com/ansible/community/issues/247

#### Upstream vendor, sanity related
- [ ] Command `bash test/ansible/sanity/integration-test-idempotent-names.sh` passes
- [ ] Command `python ./test/ansible/sanity/short-description-ends-with-period.py` passes

##### SUMMARY

**Comment by caphrim007** [comment]
PR for this is here https://github.com/ansible/ansible/pull/32281

**Comment by caphrim007** [comment]
this is now merged

**Comment by jkim711** [comment]
How do I download Ansible 2.5.0 ?
I checked https://github.com/ansible/ansible/branches/all , it only has up to 2.4.

**Comment by caphrim007** [comment]
@jkim711 ansible 2.5 is in their devel branch. You can install it via pip

`pip install --upgrade git+https://github.com/ansible/ansible.git`
